### PR TITLE
More reliable point generation in CSGObject for simple shapes

### LINKED ISSUE #24636: Explore picking interaction points in simple shapes in MonteCarloAbsorption
`MonteCarloAbsorption` picks the interaction points within the sample by choosing a random point inside the bounding box and then checking if the point is inside the actual shape. This may need a lot of random points and can fail completely if the actual shape does not fill the bounding box and the void volume is relatively large.

The geometry of simple shapes (cuboid, cylinder, ...) is well known in Mantid. It might be possible to pick the random points directly within the shapes without the detour of bounding box. This would give a usable interaction point without drawing a large number of random numbers speeding up the simulation and preventing failures.

### PR DESCRIPTION
**Pull request by soininen**
This PR improves the `CSGObject::generatePointInObject()` methods by introducing functions to find random points in simple shapes using their known geometry. Currently supported shapes are cuboid, cylinder and sphere.

The original methods picked a point in the bounding box and tested if it was within the actual shape. It was possible that none of the random points hit the shape if the shape filled the volume of the bounding box poorly (thin tilted slab, thin diagonal cylinder...). This in turn prevented algorithms like `MonteCarloAbsorption` from running. Note, that the `CSGObject::generatePointInObject()` method which takes an `activeRegion` as an argument may still fail but the probability should be lower than before since the random point needs to hit only the `activeRegion`, not the shape.

Performance tests showed that the bounding box method was actually faster for simple shapes if they filled the bounding box volume nicely whereas the specialized functions were faster for poorly filling shapes. Therefore, we
* try a few times (set to five for the moment) the bounding box method and if it fails
* pick the point using the specialized simple shape functions

For other shapes, we always fall back to the bounding box method.

**Report to:** [nobody].

**To test:**

Check that the script below runs. You can also try running it on master to see if there are any performance differences.

```python
data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = ConvertUnits(data, Target='Wavelength')
SetSample(data, Geometry={'Shape': 'FlatPlate', 'Width': 3.0, 'Height': 10.0, 'Thick':0.005,
                  'Center': [0.0,0.0,0.0], 'Angle': 135.},
                Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6', 'SampleNumberDensity': 0.07},
         )

abscor = MonteCarloAbsorption(data, Interpolation='CSpline')
```

Fixes #24636.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
@soininen Is this still in progress are is it okay to review now the windows build is fixed?

**Comment by soininen** [comment]
This is still in progress - I'd like to get #24584 merged first to use some nice utility functions there and perhaps refactor the surface meshing stuff to resemble the raster stuff there.

**Comment by martyngigg** [comment]
Cool. Thanks. I've got an idea of adding the hollow cylinder stuff to this list and someone who should have some time to look at it but it would make sense to wait until you are finished.

**Comment by soininen** [comment]
@martyngigg I am now done with the refactoring.

File: Framework/Geometry/src/Rendering/ShapeInfo.cpp [hunk 1]
```diff
@@ -34,6 +35,37 @@ void ShapeInfo::getObjectGeometry(ShapeInfo::GeometryShape &shape,
   height = m_height;
 }
 
+ShapeInfo::CuboidGeometry ShapeInfo::cuboidGeometry() const {
+  assert(m_shape == GeometryShape::CUBOID);
+  CuboidGeometry g{m_points[0], m_points[1], m_points[2], m_points[3]};
```
Comment (martyngigg) [inline Framework/Geometry/src/Rendering/ShapeInfo.cpp]: I think we can do without the temporary as this is just aggregate initialization so:
```c++
 return {m_points[0], m_points[1], m_points[2], m_points[3]};
```
directly?

File: Framework/Geometry/src/Rendering/ShapeInfo.cpp [hunk 2]
```diff
@@ -34,6 +35,37 @@ void ShapeInfo::getObjectGeometry(ShapeInfo::GeometryShape &shape,
...
+  return g;
+}
+
+ShapeInfo::HexahedronGeometry ShapeInfo::hexahedronGeometry() const {
+  assert(m_shape == GeometryShape::HEXAHEDRON);
+  HexahedronGeometry g{m_points[0], m_points[1], m_points[2], m_points[3],
```
Comment (martyngigg) [inline Framework/Geometry/src/Rendering/ShapeInfo.cpp]: Similar comment to above.

File: Framework/Geometry/src/Rendering/ShapeInfo.cpp [hunk 3]
```diff
@@ -34,6 +35,37 @@ void ShapeInfo::getObjectGeometry(ShapeInfo::GeometryShape &shape,
...
+                       m_points[4], m_points[5], m_points[6], m_points[7]};
+  return g;
+}
+
+ShapeInfo::SphereGeometry ShapeInfo::sphereGeometry() const {
+  assert(m_shape == GeometryShape::SPHERE);
+  SphereGeometry g{m_points.front(), m_radius};
```
Comment (martyngigg) [inline Framework/Geometry/src/Rendering/ShapeInfo.cpp]: Similar comment to above.

File: Framework/Geometry/src/Rendering/ShapeInfo.cpp [hunk 4]
```diff
@@ -34,6 +35,37 @@ void ShapeInfo::getObjectGeometry(ShapeInfo::GeometryShape &shape,
...
+  return g;
+}
+
+ShapeInfo::CylinderGeometry ShapeInfo::cylinderGeometry() const {
+  assert(m_shape == GeometryShape::CYLINDER);
+  CylinderGeometry g{m_points.front(), m_points.back(), m_radius, m_height};
```
Comment (martyngigg) [inline Framework/Geometry/src/Rendering/ShapeInfo.cpp]: Similar comment to above.

File: Framework/Geometry/src/Rendering/ShapeInfo.cpp [hunk 5]
```diff
@@ -34,6 +35,37 @@ void ShapeInfo::getObjectGeometry(ShapeInfo::GeometryShape &shape,
...
+  return g;
+}
+
+ShapeInfo::ConeGeometry ShapeInfo::coneGeometry() const {
+  assert(m_shape == GeometryShape::CONE);
+  ConeGeometry g{m_points.front(), m_points.back(), m_radius, m_height};
+  return g;
```
Comment (martyngigg) [inline Framework/Geometry/src/Rendering/ShapeInfo.cpp]: Similar comment to above.

File: Framework/Kernel/inc/MantidKernel/V3D.h [hunk 6]
```diff
@@ -28,10 +28,11 @@ Class for 3D vectors.
 @author Laurent C Chapon, ISIS, RAL
 @date 09/10/2007
 */
-class MANTID_KERNEL_DLL V3D {
+class MANTID_KERNEL_DLL V3D final {
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/V3D.h]: Nice improvements here. I recently inlined much of `V2D` and this gave a big performance boost.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is a great set of improvements for the sampling points within basic shapes.

Just a few minor comments.

File: Framework/Geometry/src/Rendering/ShapeInfo.cpp (hunk 1 above)
Comment (soininen) [reply Framework/Geometry/src/Rendering/ShapeInfo.cpp]: Good point.

File: Framework/Kernel/inc/MantidKernel/V3D.h (hunk 6 above)
Comment (soininen) [reply Framework/Kernel/inc/MantidKernel/V3D.h]: Much of `V3D` was already inlined, so I don't expect much difference in performance. This was mostly a small modernization operation, anyway.

**Review by martyngigg — APPROVED** [review]
This looks good. The speed up is good for the test example compared to master. The test on my debug build of master didn't finish as I got bored but with these change it finished in 30 seconds. :+1:

**Comment by peterfpeterson** [comment]
`ISIS_WISHDiffractionFocussing.WISHDiffractionFocussingAnalysisTest` is unrelated and fixed in #24829.

**Comment by DanNixon** [comment]
This seems to have caused a regression in [KernelTest.V3DTestPerformance.testElementAccessOperator](http://builds.mantidproject.org/job/master_performancetests2/Master_20branch_20performance_20tests/KernelTest.V3DTestPerformance.testElementAccessOperator.htm) (I did check for changes in the test parameters this time).

**Comment by soininen** [comment]
@DanNixon Thanks for the heads up! I'll have a look what happened there.

**Comment by soininen** [comment]
@DanNixon Can you check the *overall* `V3D` performance? I can see that there is a regression in `testElementAccess` but when I run all the `V3D` performance tests I get 2x speedup compared to `ef15643`. I increased `m_sampleSize` to 1000000000 to get a bit better statistics.

**Pull request merged**